// ngx-ourpalm-table/lib/model/ourpalm-table.js
import { OurpalmTableColumn } from "./ourpalm-table-column";
var OurpalmTable = (function () {
    function OurpalmTable(table) {
        this.originalColumns = [];
        this.tableClass = 'table table-bordered table-striped table-hover text-center';
        this.autoLoadData = true;
        this.columns = [];
        this.pagination = true;
        this.singleSelect = false;
        this.serverSort = true;
        this.multiSort = true;
        this.pageList = [10, 20, 30, 40, 50];
        this.skipPage = true;
        this.cacheKey = '';
        this.cachePageSize = false;
        this.cacheColumns = false;
        this.pagePosition = 'bottom';
        this.showRefreshBtn = true;
        this.showSettingBtn = true;
        this.checkOnSelect = true;
        this.selectOnCheck = true;
        this.ctrlSelect = false;
        this.rowViewShowType = 'column';
        this.loadData = function () {
            return null;
        };
        this.trackByFun = function (rowIndex, rowData) { return rowIndex; };
        this.openSettings = false;
        this.total = 0;
        this._rows = [];
        this._currentPage = 1;
        this._pageSize = 10;
        this.tableRows = [];
        Object.assign(this, table);
        this.changeColumns(this.columns);
        this._reloadCachePageSize();
    }
    OurpalmTable.prototype.onLoadSuccess = function (_page) {
        var page = {
            currentPage: _page.currentPage,
            pageSize: _page.pageSize,
            total: _page.total,
            rows: (_page.rows || []).slice()
        };
        this.pageSize = page.pageSize || this.pageSize;
        this.total = page.total;
        this.rows = page.rows;
        this.currentPage = page.currentPage || this.currentPage;
    };
    Object.defineProperty(OurpalmTable.prototype, "rows", {
        get: function () {
            return this._rows;
        },
        set: function (rows) {
            rows = rows || [];
            this._rows = rows;
            var __rows = rows.map(function (row, index) {
                return {
                    index: index,
                    selected: false,
                    checked: false,
                    data: row
                };
            });
            this.tableRows = __rows;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OurpalmTable.prototype, "currentPage", {
        get: function () {
            return this._currentPage;
        },
        set: function (currentPage) {
            if (this._currentPage != currentPage) {
                this._currentPage = currentPage;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OurpalmTable.prototype, "pageSize", {
        get: function () {
            return this._pageSize;
        },
        set: function (pageSize) {
            if (this._pageSize != pageSize) {
                this._pageSize = pageSize;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OurpalmTable.prototype, "allPage", {
        get: function () {
            return Math.max(Math.ceil(this.total / this.pageSize), 0);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OurpalmTable.prototype, "start", {
        get: function () {
            if (this.allPage > 0)
                return (this.currentPage - 1) * this.pageSize + 1;
            return 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OurpalmTable.prototype, "end", {
        get: function () {
            if (this.allPage > 0)
                return this.start + this.rows.length - 1;
            return 0;
        },
        enumerable: true,
        configurable: true
    });
    OurpalmTable.prototype.getOptions = function () {
        return {
            currentPage: this.currentPage,
            pageSize: this.pageSize,
            pagination: this.pagination,
            singleSelect: this.singleSelect,
            serverSort: this.serverSort,
            pageList: [].concat(this.pageList),
            skipPage: this.skipPage,
            cacheKey: this.cacheKey,
            cachePageSize: this.cachePageSize,
            cacheColumns: this.cacheColumns,
            pagePosition: this.pagePosition,
        };
    };
    OurpalmTable.prototype.getDisplayedColumns = function () {
        var columns = [];
        this.columns.forEach(function (column) {
            column.show && columns.push(Object.assign({}, column));
        });
        return columns;
    };
    OurpalmTable.prototype.getDisplayedRows = function () {
        return this.rows.map(function (row) { return Object.assign({}, row); });
    };
    OurpalmTable.prototype.getSelectedRows = function () {
        return this.tableRows.filter(function (row) { return row.selected; }).map(function (row) { return Object.assign({}, row.data); });
    };
    OurpalmTable.prototype.getCheckedRows = function () {
        return this.tableRows.filter(function (row) { return row.checked; }).map(function (row) { return Object.assign({}, row.data); });
    };
    OurpalmTable.prototype.getSortColumns = function () {
        return this.columns.filter(function (column) { return column.sort; }).map(function (col) { return Object.assign({}, col); });
    };
    OurpalmTable.prototype.changeColumns = function (columns, localStorageType) {
        var _this = this;
        if (localStorageType === void 0) { localStorageType = 'read'; }
        this.originalColumns = columns.map(function (column) { return new OurpalmTableColumn(column); });
        this.columns = columns.map(function (column) { return Object.assign(column, new OurpalmTableColumn(column)); });
        if (this.cacheKey && this.cacheColumns && window.localStorage) {
            if (localStorageType === 'write') {
                var columnArr_1 = [];
                this.columns.forEach(function (column) {
                    columnArr_1.push({ field: column.field, show: column.show });
                });
                window.localStorage.setItem("ngx-ourpalm-table-" + this.cacheKey + "-columns", JSON.stringify(columnArr_1));
            }
            else if (localStorageType === 'read') {
                var cache = window.localStorage.getItem("ngx-ourpalm-table-" + this.cacheKey + "-columns");
                if (cache) {
                    var columnArr = JSON.parse(cache);
                    if (columnArr.length == this.columns.length) {
                        var tmpColumns_1 = [];
                        columnArr.forEach((function (col1) {
                            _this.columns.forEach(function (col2) {
                                if (col1.field == col2.field) {
                                    tmpColumns_1.push(Object.assign(col2, col1));
                                }
                            });
                        }));
                        this.columns.splice(0);
                        tmpColumns_1.forEach(function (col) {
                            _this.columns.push(col);
                        });
                    }
                    else {
                        window.localStorage.removeItem("ngx-ourpalm-table-" + this.cacheKey + "-columns");
                    }
                }
            }
        }
    };
    OurpalmTable.prototype.changePageSize = function (pageSize) {
        this.pageSize = pageSize;
        this._currentPage = 1;
        this.invokeLoadData();
        if (this.cacheKey && this.cachePageSize && window.localStorage) {
            window.localStorage.setItem("ngx-ourpalm-table-" + this.cacheKey + "-pagesize", "" + pageSize);
        }
    };
    OurpalmTable.prototype.firstPage = function () {
        this.currentPage = 1;
        this.invokeLoadData();
    };
    OurpalmTable.prototype.prePage = function () {
        this.currentPage--;
        this.invokeLoadData();
    };
    OurpalmTable.prototype.nextPage = function () {
        this.currentPage++;
        this.invokeLoadData();
    };
    OurpalmTable.prototype.lastPage = function () {
        this.currentPage = this.allPage;
        this.invokeLoadData();
    };
    OurpalmTable.prototype.refresh = function () {
        this.invokeLoadData();
    };
    OurpalmTable.prototype.gotoSkipPage = function (page) {
        this.currentPage = page;
        this.invokeLoadData();
    };
    OurpalmTable.prototype.setPageData = function (page) {
        this.onLoadSuccess(page);
    };
    OurpalmTable.prototype.setOptions = function (table) {
        Object.assign(this, table);
        if (table && table.columns) {
            this.originalColumns = this.columns.map(function (column) { return new OurpalmTableColumn(column); });
            this.columns = this.columns.map(function (column) { return new OurpalmTableColumn(column); });
        }
        if (this.autoLoadData) {
            this.invokeLoadData();
        }
        this.changeColumns(this.columns);
        this._reloadCachePageSize();
    };
    OurpalmTable.prototype.checkAll = function () {
        this.tableRows = this.tableRows.map(function (row) {
            return Object.assign({}, row, { checked: true });
        });
    };
    OurpalmTable.prototype.uncheckAll = function () {
        this.tableRows = this.tableRows.map(function (row) {
            return Object.assign({}, row, { checked: false });
        });
    };
    OurpalmTable.prototype.checkRow = function (index) {
        var _this = this;
        this.tableRows = this.tableRows.map(function (row, _index) {
            if (index === _index) {
                if (_this.checkOnSelect) {
                    return Object.assign({}, row, { checked: true, selected: true });
                }
                else {
                    return Object.assign({}, row, { checked: true });
                }
            }
            return row;
        });
    };
    OurpalmTable.prototype.uncheckRow = function (index) {
        var _this = this;
        this.tableRows = this.tableRows.map(function (row, _index) {
            if (index === _index) {
                if (_this.checkOnSelect) {
                    return Object.assign({}, row, { checked: false, selected: false });
                }
                else {
                    return Object.assign({}, row, { checked: false });
                }
            }
            return row;
        });
    };
    OurpalmTable.prototype.openSetting = function () {
        this.openSettings = true;
    };
    OurpalmTable.prototype.invokeLoadData = function () {
        this.loadData(this, this.onLoadSuccess.bind(this));
    };
    OurpalmTable.prototype._reloadCachePageSize = function () {
        if (this.cacheKey && this.cachePageSize && window.localStorage) {
            var pageSize = window.localStorage.getItem("ngx-ourpalm-table-" + this.cacheKey + "-pagesize");
            if (pageSize) {
                this.pageSize = +pageSize;
            }
        }
    };
    return OurpalmTable;
}());
export { OurpalmTable };
