// ngx-ourpalm-table/lib/model/ourpalm-table-column.js
var OurpalmTableColumn = (function () {
    function OurpalmTableColumn(column) {
        this.header = '';
        this.field = '';
        this.sort = false;
        this.sortOrder = null;
        this.rownumbers = false;
        this.show = true;
        this.checkbox = false;
        this.disabledContextMenu = false;
        this.formatter = function (value, row) { return value; };
        this.sorter = function (column, row1, row2) {
            var param1 = row1[column.field];
            var param2 = row2[column.field];
            if (typeof param1 == "string" && typeof param2 == "string") {
                return param1.localeCompare(param2);
            }
            if (typeof param1 == "number" && typeof param2 == "string") {
                return -1;
            }
            if (typeof param1 == "string" && typeof param2 == "number") {
                return 1;
            }
            if (typeof param1 == "number" && typeof param2 == "number") {
                if (param1 > param2)
                    return 1;
                if (param1 == param2)
                    return 0;
                if (param1 < param2)
                    return -1;
            }
        };
        Object.assign(this, column);
    }
    return OurpalmTableColumn;
}());
export { OurpalmTableColumn };
