// glowworm/lib/confirm/confirm.module.js
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { GwConfirmComponent } from "./confirm.component";
import { GwConfirmDirective } from "./confirm.directive";
import { GwConfirmService } from "./confirm.service";
var GwConfirmModule = (function () {
    function GwConfirmModule() {
    }
    GwConfirmModule.forRoot = function () {
        return {
            ngModule: GwConfirmModule,
            providers: [GwConfirmService]
        };
    };
    return GwConfirmModule;
}());
export { GwConfirmModule };
GwConfirmModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    GwConfirmComponent,
                    GwConfirmDirective
                ],
                exports: [
                    GwConfirmDirective
                ],
                entryComponents: [
                    GwConfirmComponent
                ]
            },] },
];
GwConfirmModule.ctorParameters = function () { return []; };
