// ngx-ourpalm-table/lib/component/ourpalm-table.component.js
import { Component, ContentChild, ContentChildren, Input } from "@angular/core";
import { OurpalmTableRowViewComponent } from "./body/ourpalm-table-rowview.component";
import { OurpalmTableColumnComponent } from "./body/ourpalm-table-column.component";
import { OurpalmTableColumn } from "../model/ourpalm-table-column";
import { sortColumns } from "../utils/column-helpers";
var OurpalmTableComponent = (function () {
    function OurpalmTableComponent() {
    }
    OurpalmTableComponent.prototype.ngAfterContentInit = function () {
        var staticColumns = this.columnComponents.toArray().map(function (columnComponent) {
            var column = columnComponent.column;
            column.template = columnComponent.template;
            return new OurpalmTableColumn(column);
        });
        if (staticColumns.length > 0) {
            this.table.columns = staticColumns;
        }
        if (this.rowViewComponent) {
            this.table.rowViewTemplate = this.rowViewComponent.template;
        }
        if (this.table.autoLoadData) {
            this.table.invokeLoadData();
        }
    };
    OurpalmTableComponent.prototype.onClickRowEvent = function (row) {
        this.table && this.table.onClickRow && this.table.onClickRow(row.index, row.data);
    };
    OurpalmTableComponent.prototype.onDblClickRowEvent = function (row) {
        this.table && this.table.onDbClickRow && this.table.onDbClickRow(row.index, row.data);
    };
    OurpalmTableComponent.prototype.onClickCellEvent = function (cell) {
        this.table && this.table.onClickCell && this.table.onClickCell(cell.row.index, cell.index, cell.row.data, cell.column);
    };
    OurpalmTableComponent.prototype.onDblClickCellEvent = function (cell) {
        this.table && this.table.onDbClickCell && this.table.onDbClickCell(cell.row.index, cell.index, cell.row.data, cell.column);
    };
    OurpalmTableComponent.prototype.onRowCheckBoxChangeEvent = function (row) {
        this.table && this.table.onRowCheckBoxChange && this.table.onRowCheckBoxChange(row.data, row.index);
    };
    OurpalmTableComponent.prototype.onHeaderCheckBoxChangeEvent = function () {
        this.table && this.table.onHeaderCheckBoxChange && this.table.onHeaderCheckBoxChange();
    };
    OurpalmTableComponent.prototype.onSortColumnEvent = function (column) {
        sortColumns(column, this.table);
    };
    OurpalmTableComponent.prototype.onPagingChangeEvent = function () {
        this.table.invokeLoadData();
    };
    OurpalmTableComponent.prototype.onPagingRefreshEvent = function () {
        this.table.invokeLoadData();
    };
    return OurpalmTableComponent;
}());
export { OurpalmTableComponent };
OurpalmTableComponent.decorators = [
    { type: Component, args: [{
                selector: 'ourpalm-table',
                template: "\n        <ourpalm-table-wrapper\n                [tableClass]=\"table?.tableClass\"\n                [singleSelect]=\"table?.singleSelect\"\n                [checkOnSelect]=\"table?.checkOnSelect\"\n                [selectOnCheck]=\"table?.selectOnCheck\"\n                [ctrlSelect]=\"table?.ctrlSelect\"\n                [rowMenus]=\"table?.rowMenus\"\n\n                [cacheKey]=\"table?.cacheKey\"\n                [cacheColumns]=\"table?.cacheColumns\"\n                [cachePageSize]=\"table?.cachePageSize\"\n\n                [originalColumns]=\"table?.originalColumns\"\n                [(columns)]=\"table.columns\"\n                (onHeaderCheckBoxChange)=\"onHeaderCheckBoxChangeEvent()\"\n                (onSortColumn)=\"onSortColumnEvent($event)\"\n\n                [rows]=\"table?.tableRows\"\n                [rowView]=\"table?.rowView\"\n                [rowViewShowType]=\"table?.rowViewShowType\"\n                [rowViewTemplate]=\"table?.rowViewTemplate\"\n                (onClickRow)=\"onClickRowEvent($event)\"\n                (onDbClickRow)=\"onDblClickRowEvent($event)\"\n                (onClickCell)=\"onClickCellEvent($event)\"\n                (onDbClickCell)=\"onDblClickCellEvent($event)\"\n                (onRowCheckBoxChange)=\"onRowCheckBoxChangeEvent($event)\"\n\n                [pagination]=\"table?.pagination\"\n                [pagePosition]=\"table?.pagePosition\"\n                [(currentPage)]=\"table.currentPage\"\n                [(pageSize)]=\"table.pageSize\"\n                [total]=\"table?.total\"\n                [skipPage]=\"table?.skipPage\"\n                [pageList]=\"table?.pageList\"\n                [showRefreshBtn]=\"table?.showRefreshBtn\"\n                [showSettingBtn]=\"table.showSettingBtn\"\n                [(openSettings)]=\"table.openSettings\"\n                (onPagingChange)=\"onPagingChangeEvent()\"\n                (onPagingRefresh)=\"onPagingRefreshEvent()\">\n        </ourpalm-table-wrapper>\n    "
            },] },
];
OurpalmTableComponent.ctorParameters = function () { return []; };
OurpalmTableComponent.propDecorators = {
    'table': [{ type: Input, args: ['table',] },],
    'columnComponents': [{ type: ContentChildren, args: [OurpalmTableColumnComponent,] },],
    'rowViewComponent': [{ type: ContentChild, args: [OurpalmTableRowViewComponent,] },],
};
