// glowworm/lib/context-menu/context-menu.directive.js
import { Directive, HostListener, Input } from "@angular/core";
import { ContextMenuService } from "./context-menu.service";
var GwContextMenuDirective = (function () {
    function GwContextMenuDirective(service) {
        this.service = service;
    }
    GwContextMenuDirective.prototype.ngOnInit = function () {
        this.service.onMenuDirectiveInit();
    };
    GwContextMenuDirective.prototype.ngOnDestroy = function () {
        this.service.onMenuDirectiveDestroy();
    };
    GwContextMenuDirective.prototype.onContextMenu = function (event) {
        this.service.show.next({
            menus: this.menus,
            event: event,
            class: this.customClass
        });
        event.preventDefault();
        event.stopPropagation();
    };
    return GwContextMenuDirective;
}());
export { GwContextMenuDirective };
GwContextMenuDirective.decorators = [
    { type: Directive, args: [{
                selector: '[gw-contextmenu]'
            },] },
];
GwContextMenuDirective.ctorParameters = function () { return [
    { type: ContextMenuService, },
]; };
GwContextMenuDirective.propDecorators = {
    'menus': [{ type: Input },],
    'customClass': [{ type: Input },],
    'onContextMenu': [{ type: HostListener, args: ['contextmenu', ['$event'],] },],
};
