// glowworm/lib/context-menu/context-menu.module.js
import { NgModule } from "@angular/core";
import { GwContextMenuContainerComponent } from "./context-menu.component";
import { GwContextMenuDirective } from "./context-menu.directive";
import { CommonModule } from "@angular/common";
import { ContextMenuService } from "./context-menu.service";
import { GwCoreModule } from "../core";
var GwContextMenuModule = (function () {
    function GwContextMenuModule() {
    }
    GwContextMenuModule.forRoot = function () {
        return {
            ngModule: GwContextMenuModule,
            providers: [ContextMenuService]
        };
    };
    return GwContextMenuModule;
}());
export { GwContextMenuModule };
GwContextMenuModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    GwCoreModule.forRoot()
                ],
                declarations: [
                    GwContextMenuContainerComponent,
                    GwContextMenuDirective
                ],
                exports: [
                    GwContextMenuDirective
                ],
                entryComponents: [
                    GwContextMenuContainerComponent
                ]
            },] },
];
GwContextMenuModule.ctorParameters = function () { return []; };
