// glowworm/lib/tabs/tab.component.js
import { Component, ContentChild, Input } from "@angular/core";
var GwTabComponent = (function () {
    function GwTabComponent() {
        this.isFirstSelected = true;
    }
    GwTabComponent.prototype.ngOnInit = function () {
        this.title = this.title || this._title;
        this.content = this.content || this._content;
    };
    return GwTabComponent;
}());
export { GwTabComponent };
GwTabComponent.decorators = [
    { type: Component, args: [{
                selector: 'gw-tab',
                template: " "
            },] },
];
GwTabComponent.ctorParameters = function () { return []; };
GwTabComponent.propDecorators = {
    'tabId': [{ type: Input },],
    'title': [{ type: Input },],
    'content': [{ type: Input },],
    'lazy': [{ type: Input },],
    'disabled': [{ type: Input },],
    'closable': [{ type: Input },],
    'selected': [{ type: Input },],
    '_title': [{ type: ContentChild, args: ['title',] },],
    '_content': [{ type: ContentChild, args: ['content',] },],
};
