// glowworm/lib/confirm/confirm.directive.js
import { Directive, ElementRef, EventEmitter, HostListener, Input, Output } from "@angular/core";
import { ComponentLoaderService } from "../core/component-loader.service";
import { GwConfirmComponent } from "./confirm.component";
var GwConfirmDirective = (function () {
    function GwConfirmDirective(componentLoader, el) {
        this.componentLoader = componentLoader;
        this.el = el;
        this.zIndex = 10001;
        this.confirmText = '确认';
        this.cancelText = '取消';
        this.onConfirm = new EventEmitter();
        this.onCancel = new EventEmitter();
    }
    GwConfirmDirective.prototype.onclick = function () {
        if (!this.componentRef) {
            this.createComponent();
        }
    };
    GwConfirmDirective.prototype.createComponent = function () {
        var _this = this;
        setTimeout(function () {
            _this.componentRef = _this.componentLoader.appendComponentToBody(GwConfirmComponent);
            var confirm = _this.componentRef.instance;
            confirm.title = _this.title;
            confirm.content = _this.content;
            confirm.confirmClass = _this.confirmClass;
            confirm.confirmText = _this.confirmText;
            confirm.zIndex = _this.zIndex;
            confirm.cancelText = _this.cancelText;
            confirm.onConfirm = _this.onConfirm;
            confirm.onCancel = _this.onCancel;
            confirm.onConfirm.subscribe(function () {
                _this.ngOnDestroy();
            });
            confirm.onCancel.subscribe(function () {
                _this.ngOnDestroy();
            });
        });
    };
    GwConfirmDirective.prototype.ngOnDestroy = function () {
        if (this.componentRef) {
            this.componentLoader.removeComponentFormBody(this.componentRef);
            this.componentRef = null;
        }
    };
    return GwConfirmDirective;
}());
export { GwConfirmDirective };
GwConfirmDirective.decorators = [
    { type: Directive, args: [{
                selector: '[gw-confirm]'
            },] },
];
GwConfirmDirective.ctorParameters = function () { return [
    { type: ComponentLoaderService, },
    { type: ElementRef, },
]; };
GwConfirmDirective.propDecorators = {
    'title': [{ type: Input },],
    'content': [{ type: Input },],
    'confirmClass': [{ type: Input },],
    'zIndex': [{ type: Input },],
    'confirmText': [{ type: Input },],
    'cancelText': [{ type: Input },],
    'onConfirm': [{ type: Output },],
    'onCancel': [{ type: Output },],
    'onclick': [{ type: HostListener, args: ['click',] },],
};
