// ngx-ourpalm-table/lib/component/header/ourpalm-table-header-checkboxcell.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from "@angular/core";
var OurpalmTableHeaderCheckboxCellComponent = (function () {
    function OurpalmTableHeaderCheckboxCellComponent() {
        this.checkAll = false;
        this.checkAllChange = new EventEmitter();
        this.onHeaderCheckBoxChange = new EventEmitter();
    }
    return OurpalmTableHeaderCheckboxCellComponent;
}());
export { OurpalmTableHeaderCheckboxCellComponent };
OurpalmTableHeaderCheckboxCellComponent.decorators = [
    { type: Component, args: [{
                selector: 'ourpalm-table-header-checkobxcell',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: "\n        <!-- checkbox\u5217 -->\n        <input type=\"checkbox\"\n               [(ngModel)]=\"checkAll\"\n               (change)=\"onHeaderCheckBoxChange.emit($event);checkAllChange.emit(checkAll)\"\n        />\n    "
            },] },
];
OurpalmTableHeaderCheckboxCellComponent.ctorParameters = function () { return []; };
OurpalmTableHeaderCheckboxCellComponent.propDecorators = {
    'checkAll': [{ type: Input },],
    'checkAllChange': [{ type: Output },],
    'onHeaderCheckBoxChange': [{ type: Output },],
};
