// ngx-ourpalm-table/lib/component/header/ourpalm-table-header-cell.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from "@angular/core";
var OurpalmTableHeaderCellComponent = (function () {
    function OurpalmTableHeaderCellComponent() {
        this.onSortColumn = new EventEmitter();
    }
    return OurpalmTableHeaderCellComponent;
}());
export { OurpalmTableHeaderCellComponent };
OurpalmTableHeaderCellComponent.decorators = [
    { type: Component, args: [{
                selector: 'ourpalm-table-header-cell',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: "\n        <!-- \u6392\u5E8F\u5217 -->\n        <span *ngIf=\"column.sort\" (click)=\"onSortColumn.next(column)\" [class.sortable]=\"column.sort\">\n                {{column.header}}\n                <i class=\"fa\"\n                   [ngClass]=\"{'fa-sort-asc': column.sortOrder == 'asc', 'fa-sort-desc': column.sortOrder == 'desc', 'fa-sort': !column.sortOrder}\"></i>\n        </span>\n        <!-- \u6B63\u5E38\u5217 | \u5E8F\u53F7\u5217-->\n        <ng-container *ngIf=\"!column.sort && !column.checkbox\">\n            {{column.header}}\n        </ng-container>\n    "
            },] },
];
OurpalmTableHeaderCellComponent.ctorParameters = function () { return []; };
OurpalmTableHeaderCellComponent.propDecorators = {
    'column': [{ type: Input },],
    'onSortColumn': [{ type: Output },],
};
