// ngx-ourpalm-table/lib/component/header/ourpalm-table-header.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewEncapsulation } from "@angular/core";
var OurpalmTableHeaderComponent = (function () {
    function OurpalmTableHeaderComponent() {
        this.checkAll = false;
        this.checkAllChange = new EventEmitter();
        this.onHeaderCheckBoxChange = new EventEmitter();
        this.onSortColumn = new EventEmitter();
    }
    return OurpalmTableHeaderComponent;
}());
export { OurpalmTableHeaderComponent };
OurpalmTableHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: '[ourpalm-table-header]',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: "\n        <th *ngFor=\"let column of columns\" [class.hidden]=\"!column.show\"><!-- \u9690\u85CF\u5217 -->\n            <ng-container *ngIf=\"column.checkbox\">\n                <ourpalm-table-header-checkobxcell\n                        [checkAll]=\"checkAll\"\n                        (checkAllChange)=\"checkAllChange.emit($event)\"\n                        (onHeaderCheckBoxChange)=\"onHeaderCheckBoxChange.emit($event)\">\n                </ourpalm-table-header-checkobxcell>\n            </ng-container>\n            <ng-container *ngIf=\"!column.checkbox\">\n                <ourpalm-table-header-cell\n                        [column]=\"column\"\n                        (onSortColumn)=\"onSortColumn.emit($event)\">\n                </ourpalm-table-header-cell>\n            </ng-container>\n        </th>\n    "
            },] },
];
OurpalmTableHeaderComponent.ctorParameters = function () { return []; };
OurpalmTableHeaderComponent.propDecorators = {
    'columns': [{ type: Input },],
    'checkAll': [{ type: Input },],
    'checkAllChange': [{ type: Output },],
    'onHeaderCheckBoxChange': [{ type: Output },],
    'onSortColumn': [{ type: Output },],
};
