// ngx-ourpalm-table/lib/component/body/ourpalm-table-body-checkboxcell.component.js
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, Output, ViewEncapsulation } from "@angular/core";
var OurpalmTableBodyCheckboxCellComponent = (function () {
    function OurpalmTableBodyCheckboxCellComponent(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.onRowCheckBoxChange = new EventEmitter();
    }
    OurpalmTableBodyCheckboxCellComponent.prototype.onClickCheckBox = function (event) {
        event.stopPropagation();
    };
    OurpalmTableBodyCheckboxCellComponent.prototype.ngOnChanges = function () {
        if (this.row) {
            this._pre_checked = this.row.checked;
        }
    };
    OurpalmTableBodyCheckboxCellComponent.prototype.ngDoCheck = function () {
        if (this.row && (this.row.checked != this._pre_checked)) {
            this._pre_checked = this.row.checked;
            this.changeDetectorRef.markForCheck();
        }
    };
    return OurpalmTableBodyCheckboxCellComponent;
}());
export { OurpalmTableBodyCheckboxCellComponent };
OurpalmTableBodyCheckboxCellComponent.decorators = [
    { type: Component, args: [{
                selector: '[ourpalm-table-body-checkboxcell]',
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                template: "\n        <input type=\"checkbox\"\n               [(ngModel)]=\"row.checked\"\n               (change)=\"onRowCheckBoxChange.emit($event)\"\n               (click)=\"onClickCheckBox($event)\"\n        />\n    "
            },] },
];
OurpalmTableBodyCheckboxCellComponent.ctorParameters = function () { return [
    { type: ChangeDetectorRef, },
]; };
OurpalmTableBodyCheckboxCellComponent.propDecorators = {
    'row': [{ type: Input },],
    'onRowCheckBoxChange': [{ type: Output },],
};
